### PR TITLE
[SPARK-19276][CORE] Fetch Failure handling robust to user error handling

### PR DESCRIPTION
**Pull request by squito**
## What changes were proposed in this pull request?

Fault-tolerance in spark requires special handling of shuffle fetch
failures.  The Executor would catch FetchFailedException and send a
special msg back to the driver.

However, intervening user code could intercept that exception, and wrap
it with something else.  This even happens in SparkSQL.  So rather than
checking the thrown exception only, we'll store the fetch failure directly
in the TaskContext, where users can't touch it.

## How was this patch tested?

Added a test case which failed before the fix.  Full test suite via jenkins.

### CODE REVIEW

**Comment by squito** [comment]
cc @kayousterhout @markhamstra @mateiz 

This isn't just protecting against crazy user code -- I've seen users hit this with spark sql (because of https://github.com/apache/spark/blob/278fa1eb305220a85c816c948932d6af8fa619aa/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala#L214), so it seems important to fix.

I attempted to write a larger integration test, which reproduced the issue in a "local-cluster" setup, but got stuck.  ShuffleBlockFetcherIterator does _some_ fetches on construction, before its used as an iterator wrapped in user code.  So if the failures happen during that initialization, everything was fine before.  The failure has to happen inside the call to `shuffleBlockFetcherIterator.next()` when its called by the user's iterator for the error to happen.  I eventually was able to reproduce it with this https://github.com/squito/spark/commit/c2d27d10f32edf70e78d849967f7b7bf51495c4e but it involved hacking internals and didn't seem easy to get into a test.  I settled for a simpler unit test just on `Executor`, but open to more suggestions.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 1]
```diff
@@ -405,6 +415,13 @@ private[spark] class Executor(
           setTaskFinishedAndClearInterruptStatus()
           execBackend.statusUpdate(taskId, TaskState.FAILED, ser.serialize(reason))
 
+        case t: Throwable if task.context.fetchFailed.isDefined =>
+          // tbere was a fetch failure in the task, but some user code wrapped that exception
+          // and threw something else.  Regardless, we treat it as a fetch failure.
+          val reason = task.context.fetchFailed.get.toTaskFailedReason
+          setTaskFinishedAndClearInterruptStatus()
+          execBackend.statusUpdate(taskId, TaskState.FAILED, ser.serialize(reason))
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: nit: Probably log a similar message as above ?

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 2]
```diff
@@ -405,6 +415,13 @@ private[spark] class Executor(
           setTaskFinishedAndClearInterruptStatus()
           execBackend.statusUpdate(taskId, TaskState.FAILED, ser.serialize(reason))
 
+        case t: Throwable if task.context.fetchFailed.isDefined =>
```
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: task and task.context can be null in case exception is thrown before/while deserializing task or before task is run (or initialization of context in task.run fails).
In any of these cases, the if condition here will result in NPE, and needs to be fixed.

**Review by mridulm — COMMENTED** [review]
Thanks for the patch @squito, fixing this should definitely make spark more robust.

File: core/src/main/scala/org/apache/spark/shuffle/FetchFailedException.scala [hunk 3]
```diff
@@ -45,6 +45,12 @@ private[spark] class FetchFailedException(
     this(bmAddress, shuffleId, mapId, reduceId, cause.getMessage, cause)
   }
 
+  // SPARK-19267. We set the fetch failure in the task context, so that even if there is user-code
+  // which intercepts this exception (possibly wrapping it), the Executor can still tell there was
+  // a fetch failure, and send the correct error msg back to the driver.  The TaskContext won't be
+  // defined if this is run on the driver (just in test cases) -- we can safely ignore then.
+  Option(TaskContext.get()).map(_.setFetchFailed(this))
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/shuffle/FetchFailedException.scala]: yes, good point.  I added to the docs, does it look OK?

I also considered making the call to `TaskContext.setFetchFailed` live outside of the constructor, so at each site it was created, it would have to be called -- but I thought that seemed more dangerous.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 1 above)
Comment (squito) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: do you mean the msg I added about "TID ${taskId} completed successfully though internally it encountered unrecoverable fetch failures!"?  I wouldn't think we'd want to log anything special here.  I'm trying to make this a "normal" code path.  The user is allowed to allowed to do this.  (sparksql already does.)

we could log a warning, but then this change should be accompanied by auditing the code and making sure we never do this ourselves.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 2 above)
Comment (squito) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: oh, great point!  sorry I missed that.  I've also added a test case for this as well.

**Comment by squito** [comment]
thanks for the feedback @mridulm , all good points.  I pushed an update to address some of the points, also have some follow up discussion

File: core/src/main/scala/org/apache/spark/TaskContextImpl.scala [hunk 4]
```diff
@@ -126,4 +131,10 @@ private[spark] class TaskContextImpl(
     taskMetrics.registerAccumulator(a)
   }
 
+  private[spark] override def setFetchFailed(fetchFailed: FetchFailedException): Unit = {
+    this._fetchFailed = Some(fetchFailed)
```
Comment (vanzin) [inline core/src/main/scala/org/apache/spark/TaskContextImpl.scala]: minor: `Option(fetchFailed)`

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 5]
```diff
@@ -133,6 +123,153 @@ class ExecutorSuite extends SparkFunSuite {
...
+      partition = secondRDD.partitions(0),
+      locs = Seq(),
+      outputId = 0,
+      localProperties = new Properties(),
+      serializedTaskMetrics = serializedTaskMetrics
+    )
+
+    val serTask = serializer.serialize(task)
+    val taskDescription = fakeTaskDescription(serTask)
+
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala]: nit: too many empty lines

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 6]
```diff
@@ -133,6 +123,153 @@ class ExecutorSuite extends SparkFunSuite {
...
+  private def runTaskAndGetFailReason(taskDescription: TaskDescription): TaskFailedReason = {
+    val mockBackend = mock[ExecutorBackend]
+    var executor: Executor = null
+    try {
+      executor = new Executor("id", "localhost", SparkEnv.get, userClassPath = Nil, isLocal = true)
+      // the task will be launched in a dedicated worker thread
+      executor.launchTask(mockBackend, taskDescription)
+      val startTime = System.currentTimeMillis()
+      val maxTime = startTime + 5000
+      while (executor.numRunningTasks > 0 && System.currentTimeMillis() < maxTime) {
```
Comment (vanzin) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala]: I'd use `eventually` here, or at least `System.nanoTime` instead.

**Review by vanzin — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 1 above)
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: Yes, something along those lines ...
And I agree, we should not be doing this ourselves as well.

File: core/src/main/scala/org/apache/spark/TaskContextImpl.scala [hunk 7]
```diff
@@ -56,6 +57,10 @@ private[spark] class TaskContextImpl(
   // Whether the task has failed.
   @volatile private var failed: Boolean = false
 
+  // If there was a fetch failure in the task, we store it here, to make sure user-code doesn't
+  // hide the exception.  See SPARK-19276
+  @volatile private var _fetchFailed: Option[FetchFailedException] = None
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/TaskContextImpl.scala]: minor: can you more verbosely call this _fetchFailedException, so it's more obvious that it's not a boolean variable (like the above failed variable)

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 8]
```diff
@@ -405,6 +415,13 @@ private[spark] class Executor(
           setTaskFinishedAndClearInterruptStatus()
           execBackend.statusUpdate(taskId, TaskState.FAILED, ser.serialize(reason))
 
+        case t: Throwable if hasFetchFailure =>
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: Can the above case be eliminated now?

File: core/src/main/scala/org/apache/spark/scheduler/Task.scala [hunk 9]
```diff
@@ -137,6 +132,8 @@ private[spark] abstract class Task[T](
           memoryManager.synchronized { memoryManager.notifyAll() }
         }
       } finally {
+        // though we unset the ThreadLocal here, the context itself is still queried directly
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/Task.scala]: nit: "the context member variable" instead of just "the context" (took me a min to parse this)

**Review by kayousterhout — COMMENTED** [review]
I did a super quick review of this and have some high level thoughts (will do a more thorough review pending your thoughts on the things below):

(1) Instead of this approach, did you consider walking through the exceptions (with getCause()) to see if there's a nested FetchFailure in there?  That seems simpler, with the con of missing scenarios where the user discards the initial exception entirely.  Not sure how likely that is?  The current approach is definitely more defensive towards bad user code, but I'm hesitant about the amount of added complexity.

(2) For testing, does it help to pass in a much smaller maxBytesToFetch (spark.reduce.maxSizeInFlight) to ShuffleBlockFetcherIterator to limit the size of the initial fetches, to make it easier to wrap the FetchFailed when you want to?

**Comment by squito** [comment]
> (1) Instead of this approach, did you consider walking through the exceptions (with getCause()) to see if there's a nested FetchFailure in there? That seems simpler, with the con of missing scenarios where the user discards the initial exception entirely. Not sure how likely that is? The current approach is definitely more defensive towards bad user code, but I'm hesitant about the amount of added complexity.

yeah, I considered it but chose this for exactly the reason you suggest, to be more defensive.  It just seems way too easy for a user to get this wrong.  Even assuming the user always wraps the exception, what about bad error handling in a `finally`?  That is pretty common, even in spark, eg.:

```scala
try { 
  ...
} catch {
  case t: Throwable => throw new MySpecialAppException(t)
} finally {
  someResource.close() // oops, this can throw an exception
}
```

Given the special importance of this exception, it really seems like we should be handling it in some way that bad user code can't cover it up.

> (2) For testing, does it help to pass in a much smaller maxBytesToFetch (spark.reduce.maxSizeInFlight) to ShuffleBlockFetcherIterator to limit the size of the initial fetches, to make it easier to wrap the FetchFailed when you want to?

Do you mean for writing a larger integration test?  that would probably help make the failure more likely, but I off the top my head, I don't think that will be enough to reliable trigger the real failure (which means if its ever broken, we'll just think its a flaky test, not broken functionality).  IIRC the problem isn't just having more fetches than fit in the initial requests -- you *also* need the fetch failure to occur at a specific point in the processing. I'd need to futz around with it a while again to say for sure, though.

I'd certainly like an integration test, but eventually decided the unit test I added was sufficient.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 8 above)
Comment (squito) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: yeah, as this is now, you could eliminate this -- I left it separate for now just to highlight that we can differentiate two special cases, which we could handle in a few different ways.

1) FetchFailed is thrown, and the task fails, but its not the outer-most exception

It seems clear in this case, we should fail the task with a FetchFailure.  But do we also want to log an error or something indicating bad user code?  Kinda minor, but might be a good idea.   (Suggested by @mridulm above as well, I think.)

1a) or the FetchFailed isn't part of the thrown exception at all.

As I mentioned in my response to your other question, I'd like to consider this exactly the same as (1).

2) FetchFailed is thrown, but totally swallowed so the task succeeds

Should we succeed the task, or fail it?  I don't really know how this would happen.  It seems *really* unlikely the user meant to do this.  But then again, maybe the user did?  I chose to just log an error but still succeed the task.  (@markhamstra commented about this on the jira as well.)

its pretty easy to change the code for whatever the desired behavior is, just waiting for a clear decision.

**Comment by mridulm** [comment]
Walking up a getCause tree is not reliable - finally is one of the cases where it will fail (others being catch block's ignoring it, catch-rethrow idioms resulting in other exceptions being thrown, etc).
A cleaner solution would be great, but I am not very sure of it - using a ThreadLocal variable would be another option instead of instance variable in Task; but that would be nitpicking ...

**Comment by kayousterhout** [comment]
Ok I'm convinced re: not walking up the cause tree.  I didn't think about that finally case.  I'll do another review now.

Re: larger integration test, I didn't have a particular thing in mind -- I was mentioning that in response to your comment at it was hard to trigger the failure because of that first set of blocks that's fetched before any next() calls (so I was hoping you had some integration test in mind that would leverage that).   But fine not to have one if it's still too complex to do.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 10]
```diff
@@ -405,6 +415,13 @@ private[spark] class Executor(
...
+          // tbere was a fetch failure in the task, but some user code wrapped that exception
+          // and threw something else.  Regardless, we treat it as a fetch failure.
+          val reason = task.context.fetchFailed.get.toTaskFailedReason
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: tiny nit: but does it make sense to store the taskFailedReason (rather than the actual exception) in the task context?

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 11]
```diff
@@ -133,6 +123,153 @@ class ExecutorSuite extends SparkFunSuite {
       }
     }
   }
+
+  test("SPARK-19276: Handle Fetch Failed for all intervening user code") {
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala]: how about "Handle FetchFailedExceptions that are hidden by user exceptions"?

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 12]
```diff
@@ -133,6 +123,153 @@ class ExecutorSuite extends SparkFunSuite {
...
+      outputId = 0,
+      localProperties = new Properties(),
+      serializedTaskMetrics = serializedTaskMetrics
+    )
+
+    val serTask = serializer.serialize(task)
+    val taskDescription = fakeTaskDescription(serTask)
+
+
+    val failReason = runTaskAndGetFailReason(taskDescription)
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala]: can you add a comment about what's going on here? I think the FFE gets thrown because the shuffle map data was never generated? And then you're checking that it's correctly accounted for, even though the user RDD code wrapped the exception in something else?

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 13]
```diff
@@ -133,6 +123,153 @@ class ExecutorSuite extends SparkFunSuite {
...
+    assert(failReason.isInstanceOf[FetchFailed])
+  }
+
+  test("Gracefully handle error in task deserialization") {
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala]: is this test related to this PR? (seems useful but like it should be in its own PR?)

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 14]
```diff
@@ -133,6 +123,153 @@ class ExecutorSuite extends SparkFunSuite {
...
+      .statusUpdate(meq(0L), meq(TaskState.FAILED), statusCaptor.capture())
+    // first statusUpdate for RUNNING has empty data
+    assert(statusCaptor.getAllValues().get(0).remaining() === 0)
+    // second update is more interesting
+    val failureData = statusCaptor.getAllValues.get(1)
+    SparkEnv.get.closureSerializer.newInstance().deserialize[TaskFailedReason](failureData)
+  }
+}
+
+class FakeShuffleRDD(sc: SparkContext) extends RDD[Int](sc, Nil) {
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala]: about about FetchFailureThrowingShuffleRDD? (to make it obvious what the point of this is?)

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala [hunk 15]
```diff
@@ -133,6 +123,153 @@ class ExecutorSuite extends SparkFunSuite {
...
+    failReason match {
+      case ef: ExceptionFailure =>
+        assert(ef.exception.isDefined)
+        assert(ef.exception.get.getMessage() === "failure in deserialization")
+      case _ =>
+        fail("unexpected failure type: $failReason")
+    }
+  }
+
+  private def mockEnv(conf: SparkConf, serializer: JavaSerializer): SparkEnv = {
```
Comment (kayousterhout) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala]: can this and the below method have a verb in the name, since they're doing something rather than just getters? createMockEnv?

**Review by kayousterhout — COMMENTED** [review]
The approach here looks good.  I was hoping there might be a way to simplify but I couldn't think of anything (and the commenting is helpful for noting why stuff needed to be added)

File: core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala (hunk 13 above)
Comment (squito) [inline core/src/test/scala/org/apache/spark/executor/ExecutorSuite.scala]: @mridulm pointed out this bug in an earlier version of this pr, so I fixed the bug and added a test case.  But in any case, I've separated this out into https://github.com/apache/spark/pull/16930 / https://issues.apache.org/jira/browse/SPARK-19597

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 16]
```diff
@@ -400,8 +410,16 @@ private[spark] class Executor(
...
-          val reason = ffe.toTaskFailedReason
+        case t: Throwable if hasFetchFailure =>
+          val reason = task.context.fetchFailed.get.toTaskFailedReason
+          if (!t.isInstanceOf[FetchFailedException]) {
+            // there was a fetch failure in the task, but some user code wrapped that exception
+            // and threw something else.  Regardless, we treat it as a fetch failure.
+            logWarning(s"TID ${taskId} encountered a ${classOf[FetchFailedException]} and " +
+              s"failed, but did not directly throw the ${classOf[FetchFailedException]}.  " +
+              s"Spark is still handling the fetch failure, but these exceptions should not be " +
+              s"intercepted by user code.")
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: @mridulm @kayousterhout how is this msg?  open to other suggestions.  I'm not sure exactly what to recommend to the user instead.

**Comment by squito** [comment]
Jenkins, retest this please

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 16 above)
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: I worry that this is slightly misleading because there's not necessarily anything *bad* happening here (e.g., in the SQL case), and the user-thrown exception is getting permanently lost. What about something more like 

```
 logWarning(s"TID ${taskId} encountered a ${classOf[FetchFailedException]} and " +
               s"failed, but the ${classOf[FetchFailedException]} was hidden by another " +
               s"exception: $t.  Spark is handling this like a fetch failure and ignoring the " +
               s"other exception.")
```

File: core/src/main/scala/org/apache/spark/shuffle/FetchFailedException.scala [hunk 17]
```diff
@@ -45,6 +50,12 @@ private[spark] class FetchFailedException(
     this(bmAddress, shuffleId, mapId, reduceId, cause.getMessage, cause)
   }
 
+  // SPARK-19276. We set the fetch failure in the task context, so that even if there is user-code
+  // which intercepts this exception (possibly wrapping it), the Executor can still tell there was
+  // a fetch failure, and send the correct error msg back to the driver.  The TaskContext won't be
+  // defined if this is run on the driver (just in test cases) -- we can safely ignore then.
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/shuffle/FetchFailedException.scala]: This last sentence is confusing.  A task that runs locally on the driver can still hit fetch failures right?  Or are you saying the TaskContext will only be not defined in test cases?

**Review by kayousterhout — COMMENTED** [review]
A few last things.  Also looks like this duplicates some of the functionality in #16930, which can be fixed when that's merged

File: core/src/main/scala/org/apache/spark/executor/Executor.scala (hunk 16 above)
Comment (mridulm) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: @kayousterhout While I like the message, spark sql should not be catching that exception to begin with anyway.

Btw, the impact of ignoring the exception here is needs to be also considered ... "catch Throwable" block does some interesting things for accumulator updates, isFatalError.
Atleast the latter must be handled here (an OOM being raised for example) - not sure about accumulator updates ...

Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: I agree with @mridulm that it looks like these lines (473-475 below) need to be added here: 

```
if (Utils.isFatalError(t)) {
             SparkUncaughtExceptionHandler.uncaughtException(t)		             
}
```

I'm less sure about the accumulator updates.  It looks like the old code doesn't report accumulators for fetch failed exceptions, but it's not clear to me why we'd report them for some kinds of exceptions but not others.  The simplest thing to do seems to be the current approach (since it roughly maintains the old behavior of not updating accumulators for fetch failures) but I don't have a good sense for why this is or is not correct.

Comment (squito) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: Thanks, I like that msg better.  I changed it slightly so the original exception is at the end, otherwise its hard to tell where the original exception ends and you are back to the error msg.  Here's what the new msg looks like from the test case now:

```
17/02/27 16:33:43.953 Executor task launch worker for task 0 WARN Executor: TID 0 encountered a org.apache.spark.shuffle.FetchFailedException and failed, but the org.apache.spark.shuffle.FetchFailedException was hidden by another exception.  Spark is handling this like a fetch failure and ignoring the other exception: java.lang.RuntimeException: User Exception that hides the original exception
```

You have a good point about the uncaught exception handler, I have added that back.  I wondered whether I should add those lines inside the `case t: Throwable if hasFetchFailure ` block, or make it a condition for the case itself `case t: Throwable if hasFetchFailure && !Utils.isFatalError(t)`.  I decided to make it part of the condition, since that is more like the old behavior, and a fetch failure that happens during an OOM may not be real.

I also looked into adding a unit test for this handling -- it requires some refactoring, potentially more work than its worth, so I put it in a separate commit.

 I'd rather avoid changing the behavior for accumulators here.  Accumulators have such weird semantics its not clear what they should do, we can fix that separately if we really want to.

File: core/src/main/scala/org/apache/spark/shuffle/FetchFailedException.scala (hunk 17 above)
Comment (squito) [inline core/src/main/scala/org/apache/spark/shuffle/FetchFailedException.scala]: sorry, I've reworded this.  The issue is that we have test cases where the TaskContext isn't defined, and so we'd hit an NPE without the `Option` wrapper.  But in general, the `TaskContext` should always be defined anytime we'd create a `FetchFailure`.

The alternative would be to track down the test cases w/out a `TaskContext`, and add one back.

**Comment by kayousterhout** [comment]
Jenkins retest this please (filed https://issues.apache.org/jira/browse/SPARK-19772)

**Review by kayousterhout — APPROVED** [review]
LGTM -- thanks for adding the test for fatal errors

**Comment by squito** [comment]
@mridulm look ok to you too?  I plan on merging soon.

I just made a small change to the comments (I copied and pasted incorrect comments in the last test case I added)

**Comment by mridulm** [comment]
@squito It looks good to me, thanks for the changes !

**Comment by kayousterhout** [comment]
I merged this into master.  Thanks @squito!
